model/DataObjectSchema.php

```diff
@@ -0,0 +1,377 @@
+<?php
+
+use SilverStripe\Framework\Core\Configurable;
+use SilverStripe\Framework\Core\Injectable;
+use SilverStripe\Model\FieldType\DBComposite;
+
+/**
+ * Provides dataobject and database schema mapping functionality
+ */
+class DataObjectSchema {
+	use Injectable;
+	use Configurable;
+
+	/**
+	 * Default separate for table namespaces. Can be set to any string for
+	 * databases that do not support some characters.
+	 *
+	 * Defaults to \ to to conform to 3.x convention.
+	 *
+	 * @config
+	 * @var string
+	 */
+	private static $table_namespace_separator = '\\';
+
+	/**
+	 * Cache of database fields
+	 *
+	 * @var array
+	 */
+	protected $databaseFields = [];
+
+	/**
+	 * Cache of composite database field
+	 *
+	 * @var array
+	 */
+	protected $compositeFields = [];
+
+	/**
+	 * Cache of table names
+	 *
+	 * @var array
+	 */
+	protected $tableNames = [];
+
+	/**
+	 * Clear cached table names
+	 */
+	public function reset() {
+		$this->tableNames = [];
+		$this->databaseFields = [];
+		$this->compositeFields = [];
+	}
+
+	/**
+	 * Get all table names
+	 *
+	 * @return array
+	 */
+	public function getTableNames() {
+		$this->cacheTableNames();
+		return $this->tableNames;
+	}
+
+	/**
+	 * Given a column name, build the "Table"."Column" string appropriate for querying with.
+	 * This method will correctly match the column to the data class and table.
+	 *
+	 * Note that any fixed_fields will use the base table. Use quotedID() for joining
+	 * on the ID of non-base tables.
+	 *
+	 * @param string $class Class name
+	 * @param string $column Column name
+	 * @return string The quoted column, if this column belongs to this model
+	 */
+	public function quotedColumn($class, $column) {
```

Comment:
the return values for each field could be a SQLColumn, as you said above, but we could implement __toString to return the quoted value for each. Then the below will work.

```php
$helper = $obj->getTableHelper();
return "SELECT {$helper->Title} FROM {$helper->Title->Table} INNER JOIN {$helper->BaseTable} ON {$helper->BaseTable->ID} = {$helper->ChildTable->ID}";
```